alphabase/psm_reader/sage_reader.py

```diff
@@ -13,22 +17,205 @@
 )
 
 
-def sage_spec_idx_from_scannr(scannr: str) -> int:
-    """Extract the spectrum index from the scannr field in Sage output.
+class SageModificationTranslation:
+    def __init__(
+        self,
+        custom_translation_df: pd.DataFrame = None,
+        ppm_tolerance: int = 10,
+        mp_process_num=10,
+    ):
+        """Translate Sage style modifications to alphabase style modifications.
+        By default, the translation is done by matching the observed mass and location to the UniMod database.
+        If a custom translation dataframe is provided, the translation will be done based on the custom translation dataframe first.
+
+        Parameters
+        ----------
+        custom_translation_df : pd.DataFrame
```

Comment:
How do users get this modification translation df?

Reply:
Got it..

Reply:
Ahh, still not clear after reading all code

Reply:
Maybe the jupyter notebook with the tests helps. The translation df can always be automatically generated but might be ambiguous because it only matches the mass and the AA.

In such cases you could define 
```
custom_translation_df = pd.DataFrame({
    'modification': ['[+42.010567]'],
    'matched_mod_name': ['ThisModDoesNotExist@K']
})
```

For a single mod.
Same if you don't have it in the MOD_DF

Reply:
Can we provide some common mod map in psm_reader.yaml?

Reply:
we could, alltough it will use all of MOD_DF anyways. We only need to provide a `custom_translation_df` for edge cases.

Reply:
Like what we did in pepxml. If we used MOD_DF, there will be quite a few isobaric PTMs. But I think it is fine currently, just keep in mind about this.

```
msfragger_pepxml:
  reader_type: msfragger_pepxml
  rt_unit: second
  column_mapping:
    'sequence': 'peptide'
    'charge': 'assumed_charge'
    'rt': 'retention_time_sec'
    'query_id': 'spectrum'
    'scan_num': 'start_scan'
    'score': 'expect'
    'fdr': 'expect'
    'proteins': 'protein'
    'raw_name': 'raw_name'
    'mobility': 'ion_mobility'
  modification_mapping:
    '': ''
  mass_mapped_mods:
    - 'Oxidation@M' #other Oxidation@X are not needed here
    - 'Carbamidomethyl@C'
    - 'Phospho@S' #TY are not needed here
    - 'GlyGly@K'
    - 'Cysteinyl@C'
    - 'Acetyl@Any N-term'
    - 'Glu->pyro-Glu@E^Any N-term'
    - 'Gln->pyro-Glu@Q^Any N-term'
    - 'Dimethyl@K' # Any N-term is not needed here as it will be infered in-the-fly
    - 'Methyl@E' #an example of a PTM that can be C-term
  mod_mass_tol: 0.1 # Da
```